addons/stock_account/models/product.py

```diff
@@ -615,6 +616,10 @@ def _compute_average_price(self, qty_invoiced, qty_to_invoice, stock_moves):
         if not qty_to_invoice:
             return 0.0
 
+        returned_quantities = defaultdict(float)
+        for move in stock_moves:
+            if move.origin_returned_move_id:
+                returned_quantities[move.origin_returned_move_id.id] += move.product_uom_qty
```

Comment:
move.product_qty to avoid UoM mistakes

Reply:
It's even better to take the quantity on the return layer. it's already converted in the product uom and it excludes the potential non-valuated move lines.
Also, returned move are part of `stock_moves` what happens with their valuation layers in the loop. Don't you need to bypass them with your patch ? 
Last request : adding a test would be nice to be sure we don't add bug for this flowin the future. Thanks :)